### ISSUE TITLE
Implement Zigbee in Arduino

### ISSUE THREAD

**Issue by P-R-O-C-H-Y** [espressif/arduino-esp32#10135]
### Related area

Create an API for Zigbee within Arduino

### Hardware specification

ESP32-C6 and ESP32-H2 as Standalone Nodes. Other SoC can be used as radio co-processor attached to a RPC (802.15.4 radio layer).

### Is your feature request related to a problem?

This is an issue to track the progress of developing Zigbee library, a new API for Arduino.

### Describe the solution you'd like

**Initial tasks:**
- [ ] Add zigbee-sdk to Arduino lib-builder https://github.com/espressif/esp32-arduino-lib-builder/pull/146
- [ ] Add zigbee support #8844
- [ ] Port basic IDF examples to Arduino #9024 #9627

**Zigbee library (wrapper) tasks:**
- [ ] Create Zigbee classes and implement all Zigbee roles
- [ ] Implement Zigbee network scanning
- [ ] Allow multiple endpoints on same Zigbee device (test needed)
- [ ] Implement most HA devices (lights, switches, sensors, etc..)
- [ ] Implement setting Manufacturer and model name
- [ ] Update ported examples to use the Zigbee Library
- [ ] Documentation

more tasks will be added while in progress.

### Describe alternatives you've considered

No other usable Zigbee wrapper around.

### Additional context

Related issues:
#8807 #9156 #9962 #9480

### I have checked existing list of Feature requests and the Contribution Guide

- [X] I confirm I have checked existing list of Feature requests and Contribution Guide.

**Comment by elgerg** [comment]
Hi, love the work so far. I'm so excited to get this running. It looks like it's going to be the next best thing since sliced bread!

Are there any intentions to add power source/battery level into the config?

Many thanks and appreciate the hard work!

**Comment by P-R-O-C-H-Y** [comment]
Dev update 12/08/2024 pushed to the dev branch:
- Implemented Factory reset of Zigbee device, in order to connect to new network without reflashing/erasing flash
- Implemented optional setting for Manufacturer and Model names
- Added option to allow endpoint to have multiple endpoint connected -> switch - 2 lights (tested)
- Implemented easy transfer from device it to Device type (0x0000 = ESP_ZB_HA_ON_OFF_SWITCH_DEVICE_ID -> "General On/Off switch".
- Minor sketches update

**Comment by elgerg** [comment]
Oh wow, that looks amazing! So easy, I'm going to love this. Thanks!

**Comment by P-R-O-C-H-Y** [comment]
> Hi, love the work so far. I'm so excited to get this running. It looks like it's going to be the next best thing since sliced bread!
> 
> Are there any intentions to add power source/battery level into the config?
> 
> If mains powered could it act as a router as well or is that too much for a H2? Maybe ok for the C6..
> 
> Many thanks and appreciate the hard work!

Thank you for your feedback 👍 I really appreciate it.

About the power source/baterry level I will add it on the list as good to have, so I will look after having the main part done :)

**Comment by P-R-O-C-H-Y** [comment]
Dev Update 14/08/2024 pushed to dev branch:

- Implemented Color Dimmable Light and Colour Dimmer Switch endpoints, working in RGB colour space (automatically convert to/from XYZ).
- Added examples: Zigbee_Color_Dimmable_Light, Zigbee_Color_Dimmer_Switch
- Minor code updates

**Comment by elgerg** [comment]
Nice,
Any chance you can take a look at adding temp sensor and light sensor?
They are the 2 I'm most eager to try, I'm currently using the temp sensor and it works great (even if the code is way difficult to understand).
Thanks!

**Comment by P-R-O-C-H-Y** [comment]
> Nice, Any chance you can take a look at adding temp sensor and light sensor? They are the 2 I'm most eager to try, I'm currently using the temp sensor and it works great (even if the code is way difficult to understand). Thanks!

@elgerg The temperature sensor + thermostat are on the list of upcoming tasks.
The light sensor device type is not present [in the list](https://github.com/espressif/esp-zigbee-sdk/blob/7dab668fef089bd8ccf0b137df3ed690bd66fc13/components/esp-zigbee-lib/include/zcl/esp_zigbee_zcl_common.h#L99) in esp32-zigbee-sdk, it should have value 0x106, which is not present.
Can you open a feature request to the esp-zigbee-sdk repository please? Seems that the main part is already [there](https://github.com/espressif/esp-zigbee-sdk/blob/7dab668fef089bd8ccf0b137df3ed690bd66fc13/components/esp-zigbee-lib/include/zcl/esp_zigbee_zcl_illuminance_meas.h#L38) which contains all the clusters and attributes for luminance sensor.

**Comment by elgerg** [comment]
Thanks for adding to https://github.com/espressif/esp-zigbee-sdk/issues/401
I'm quite excited about being able to create my own Temp/Lumin Zigbee sensors :D
Thanks for the continued effort!

**Comment by elgerg** [comment]
Hi @P-R-O-C-H-Y 

Any chance I can add this to the "Nice to have" list?

In the current Temp sensor version there is an identify cluster (?) added:

```
    esp_zb_cluster_list_add_identify_cluster(cluster_list, esp_zb_identify_cluster_create(&(temperature_sensor->identify_cfg)), ESP_ZB_ZCL_CLUSTER_SERVER_ROLE)

```

Any chance you can add something like that to the list so we can blink an LED when triggered? No idea how to do that at the moment..

Thanks!

**Comment by P-R-O-C-H-Y** [comment]
> Hi @P-R-O-C-H-Y
> 
> Any chance I can add this to the "Nice to have" list?
> 
> In the current Temp sensor version there is an identify cluster (?) added:
> 
> ```
>     esp_zb_cluster_list_add_identify_cluster(cluster_list, esp_zb_identify_cluster_create(&(temperature_sensor->identify_cfg)), ESP_ZB_ZCL_CLUSTER_SERVER_ROLE)
> ```
> 
> Any chance you can add something like that to the list so we can blink an LED when triggered? No idea how to do that at the moment..
> 
> Thanks!

Hi @elgerg,

Can you explain a bit more? I am not sure if I got it correctly. 
What would be "Nice to have"? Option to customise identify cluster? I don't also get the LED blink.

**Comment by elgerg** [comment]
Hi @P-R-O-C-H-Y 
I should have been more specific.

If you have a look at this:
![image](https://github.com/user-attachments/assets/6212db4f-3a9f-48d1-b383-337b852ba511)

There is an identify button. In theory this should allow for the device to identify itself by blinking an LED or something when the button is pressed so you can figure out which device is which.

This article explains really well on what identify is:
https://medium.com/@omaslyuchenko/hello-zigbee-part-22-identify-cluster-90cf12680306

So what would be good to have is a simple way of using this just like you are doing with the rest of the wrapper.

Does that make a little more sense?

Thanks

**Comment by P-R-O-C-H-Y** [comment]
> Hi @P-R-O-C-H-Y I should have been more specific.
> 
> If you have a look at this: ![image](https://private-user-images.githubusercontent.com/24628571/358548431-6212db4f-3a9f-48d1-b383-337b852ba511.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4MDM1MzYsIm5iZiI6MTcyMzgwMzIzNiwicGF0aCI6Ii8yNDYyODU3MS8zNTg1NDg0MzEtNjIxMmRiNGYtM2E5Zi00OGQxLWIzODMtMzM3Yjg1MmJhNTExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE2VDEwMTM1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJlODBiMjliZDkxMWYxMjJlNGEyOWFjNWQ1Y2U3ZjY4ZGJlOWIyZjVkZDg3NGM1ZDVmNTVjYzg5ODRkMTVjZjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tckq5zu7GjZWm7RW3jcQoAh1isPYrAMqowiYzgIWShU)
> 
> There is an identify button. In theory this should allow for the device to identify itself by blinking an LED or something when the button is pressed so you can figure out which device is which.
> 
> This article explains really well on what identify is: https://medium.com/@omaslyuchenko/hello-zigbee-part-22-identify-cluster-90cf12680306
> 
> So what would be good to have is a simple way of using this just like you are doing with the rest of the wrapper.
> 
> Does that make a little more sense?
> 
> Thanks

Thanks for explaining. That can be really helpful. Adding it to the list :)

**Comment by elgerg** [comment]
Excellent. Thanks :)

**Comment by P-R-O-C-H-Y** [comment]
Dev Update 16/08/2024 pushed to dev branch:

- Implemented Zigbee network scanning (mostly match WiFi scan API)
- Added Zigbee_Network_Scan example

**Comment by elgerg** [comment]
If you're taking requests I'd also like to implement some kind of way of getting the current time from the coordinator to display on a screen. Do you know if it's possible to get the time into some sort of callback?

Thanks :)

**Comment by P-R-O-C-H-Y** [comment]
> If you're taking requests I'd also like to implement some kind of way of getting the current time from the coordinator to display on a screen. Do you know if it's possible to get the time into some sort of callback?
> 
> Thanks :)

Do you have any example of this feature?

**Comment by P-R-O-C-H-Y** [comment]
Dev Update 28/08/2024 pushed to dev branch:

- Implemented Temperature sensor and Thermostat endpoints.
- Added examples: Zigbee_Temperature_Sensor, Zigbee_Thermostat
- Implemented configure report handler.
- Updated READMEs and description of examples.
- Minor code updates

**Comment by elgerg** [comment]
> Dev Update 28/08/2024 pushed to dev branch:
> 
> * Implemented Temperature sensor and Thermostat endpoints.
> * Added examples: Zigbee_Temperature_Sensor, Zigbee_Thermostat
> * Implemented configure report handler.
> * Updated READMEs and description of examples.
> * Minor code updates

This looks awesome!

Do you know when it'll end up in the next release?
Thanks 🙏

**Comment by Hedda** [comment]
@P-R-O-C-H-Y Tip; here is a other Arduino library for creating Zigbee Home Automation devices (with links to some examples):

* https://github.com/prairiesnpr/zha_arduino

**Comment by P-R-O-C-H-Y** [comment]
> > Dev Update 28/08/2024 pushed to dev branch:
> 
> > 
> 
> > * Implemented Temperature sensor and Thermostat endpoints.
> 
> > * Added examples: Zigbee_Temperature_Sensor, Zigbee_Thermostat
> 
> > * Implemented configure report handler.
> 
> > * Updated READMEs and description of examples.
> 
> > * Minor code updates
> 
> 
> 
> This looks awesome!
> 
> 
> 
> Do you know when it'll end up in the next release?
> 
> Thanks 🙏

I am doing the best that the first version can be released soon.

**Comment by P-R-O-C-H-Y** [comment]
Dev Update 26/09/2024 - PR ready to be reviewed.

- Refactored methods and classes (exchanged virtual methods for callback functions making it more simpler 
- Implemented Identify (tested with HomeAssistant - SONOFF USB Dongle gateway + ESP with RGB example)
- All examples got updated due to the refactor

**Comment by P-R-O-C-H-Y** [comment]
The Zigbee library got merged! Now it's available in both`master` and `release/v3.1` branches.
So in next release 3.0.6 and 3.1.0-rc2 it will be included.

**Comment by lboue** [comment]
Great job!

**Comment by Abdull** [comment]
awesome to hear!

I'm inexperienced what is required to implement software for Zigbee devices.

Having a look at directory https://github.com/espressif/arduino-esp32/tree/master/libraries/Zigbee/src/ep , do I deduce correctly that each "category" of Zigbee device type would require its own implementation files?

Maybe you can give hints how others could join?

**Comment by P-R-O-C-H-Y** [comment]
> awesome to hear!
> 
> I'm inexperienced what is required to implement software for Zigbee devices.
> 
> Having a look at directory https://github.com/espressif/arduino-esp32/tree/master/libraries/Zigbee/src/ep , do I deduce correctly that each "category" of Zigbee device type would require its own implementation files?
> 
> Maybe you can give hints how others could join?

Hello @Abdull, take a look at the basics of the Zigbee protocol. You can find many sources and videos on the internet ;)
As you noticed, Zigbee device can have 1 or more endpoints, which specify its type (light, sensor, switch, etc.). Each type need to have its own implementation as you found in the library (the link you posted). I am open on any suggestion what EPs to add in next updates and also I would be really happy to see a PR with any new EP support. As a reference you can take a look at the EPs I have added. For more details check the [esp32-zigbee-sdk](https://github.com/espressif/esp-zigbee-sdk) repository and sources there.

**Comment by elgerg** [comment]
Hi @P-R-O-C-H-Y 

If you're taking requests these are the sort of things I would like to see:

Human presence detector (PIR etc)
Luminosity/Light sensor
Battery monitoring
Sleepy device

Thanks!

**Comment by MeisterQ** [comment]
Id also like to put a feature request:

- TRV support with stuff like external Temp sensor, window open, temperature setpoint, etc (like everything the Shelly TRV supports but only on zigbee)

**Comment by P-R-O-C-H-Y** [comment]
> Id also like to put a feature request:
> 
> * TRV support with stuff like external Temp sensor, window open, temperature setpoint, etc (like everything the Shelly TRV supports but only on zigbee)

Hi @MeisterQ, the TRV is just a thermostat. We have support for this type of endpoint. Also we have a temperature sensor endpoint supported as well. Just regular sensor for example for windows, doors, etc. is not yes supported. So if you plan to build your own TRV you can already start prototyping with the stuff we already have :)

**Comment by Hedda** [comment]
> If you're taking requests these are the sort of things I would like to see:
> 
> Human presence detector (PIR etc) 
> Luminosity/Light sensor
> Time Client (get the current time from the coordinator)
> Battery monitoring 
> Sleepy device
> 
> Thanks!

FYI, there are more Zigbee examples in the [esp32-zigbee-sdk](https://github.com/espressif/esp-zigbee-sdk) repository which could probably be converted to this new API. See:

* https://github.com/espressif/esp-zigbee-sdk/tree/main/examples

That includes examples for "deep sleep" and "light sleep" (standby) modes for Zigbee End Device:

* https://github.com/espressif/esp-zigbee-sdk/tree/main/examples/esp_zigbee_sleep

As for Battery monitoring, I read note Arduino recently released an open-source Arduino Power Management library for Arduino Pro that could maybe be ported? It includes battery management and charger manager + sleep and standby mode

* https://github.com/arduino-libraries/Arduino_PowerManagement/
  * https://github.com/arduino-libraries/Arduino_PowerManagement/tree/main/examples

More on that here 

* https://www.cnx-software.com/2024/09/26/arduino-power-management-library-arduino-pro-modules/

Time Client looks to be missing, and I agree that Zigbee Time Client is needed to make a standa-alone smart thermostat or HVAC.

**Comment by MeisterQ** [comment]
> > Id also like to put a feature request:
> > 
> > * TRV support with stuff like external Temp sensor, window open, temperature setpoint, etc (like everything the Shelly TRV supports but only on zigbee)
> 
> Hi @MeisterQ, the TRV is just a thermostat. We have support for this type of endpoint. Also we have a temperature sensor endpoint supported as well. Just regular sensor for example for windows, doors, etc. is not yes supported. So if you plan to build your own TRV you can already start prototyping with the stuff we already have :)

Thank you. That could be a start. None of the TRVs existing meet my requirements

**Comment by gdivry** [comment]
Thanks a lot for this incredible work 
Can I use the esp32c6 as a Zigbee coordinator with the lib ?

**Comment by dpharris** [comment]
Could README.md files be possibly added to the arduino-esp32/libraries
/Zigbee/ directory,  to the example directory, and/or the individual
examples?

These would be very helpful to people new to this.  They could also include
actual or links to pictures and videos.

Thanks
David

**Comment by Hedda** [comment]
> > Can I use the esp32c6 as a Zigbee coordinator with the lib ?
> 
> As I understand this is currently only to use ESP32-C6/ESP32-H2 as a ”Zigbee End Device” (ZED) or ”Zigbee Router” device (ZR) stand-alone nodes (i.e. standalone devices that can be added to an existing Zigbee network and then controlled via it), not as Zigbee Coordinator or Zigbee Gateway.
> 
> If you want a all-in-one Zigbee Gateway using ESP32 then can I suggest that you look into adding support for [esp-zigbee-sdk](https://github.com/espressif/esp-zigbee-sdk) support to Tasmota's sub-project called Zigbee2Tasmota (Z2T) which is currently only compatible with Silicon Labs and Texas Instruments as Zigbee Coordinator but it might give you ideas:
> 
> * https://tasmota.github.io/docs/Zigbee/
> 
> and
> 
> * https://tasmota.github.io/docs/Zigbee-Internals/

@s-hadinger to go off on a tangent with question above; what exact functions are still missing for this new Zigbee API for Arduino to potentially be useful for a Zigbee Gateway project like Zigbee2Tasmota (Z2T)?

**Comment by s-hadinger** [comment]
> @s-hadinger to go off on a tangent with question above; what exact functions are still missing for this new Zigbee API for Arduino to potentially be useful for a Zigbee Gateway project like Zigbee2Tasmota (Z2T)?

I don't know, I have not checked. But there are no plans to port Z2T to H2/C6 due to the predominance of CC2652P and low number of devices using H2/C6. Also for the future, I expect Thread to progressively take over Zigbee. So I think we are well served with Zigbee as of today with both EFR32/CC2652P compatibility

**Comment by P-R-O-C-H-Y** [comment]
Another update for the Zigbee library -> PR #10551 

- Support for sleepy device (only deep-sleep is currently possible).
- Support for selecting power source for the device. If battery is selected, the remaining percentage can be updated.
- Support for Humidity cluster, which was added as optional for the Temperature sensor as there is no HA Device type for humidity sensor.
- New example Zigbee_Temp_Hum_Sensor_Sleepy, which demonstrates all the new functionalities
- Mirror fixes and improvements

**Comment by P-R-O-C-H-Y** [comment]
I have also did a power consumption test for the sleepy device. Sharing the picture of consumption when sleep was set to 30s, after wakeup, the connection was done, measured data and reported. Then back to sleep.

<img width="2956" alt="Screenshot 2024-10-25 at 14 44 24" src="https://github.com/user-attachments/assets/cf482667-6cab-4271-a264-a781a902d6a5">
You can see the average consumption was around 6mA. The more time the device will sleep the better power efficiency you can get.

**Comment by tlanfer** [comment]
This is great work, and i managed to get it to work immediately using a waveshare ESP32-C6 devkit board.
I'm curious about the possiblities with sleep. One of the major reasons i'm using zigbee devices is when they need to run on low power. I have a bunch TRVs that run on battery at last a long time (months), yet respond to commands almost immediately (less than 5 seconds delay). My specific usecase is having a battery powered LED light that i only need to activate every now and then, but should not drain the battery when not in use. Zigbee would be perfect for this.

With deep sleep i'm getting a full reboot, and if i want reasonably quick response times i can only sleep for relatively short times. So i would assume i need some form of light sleep for that? Just wake up enough to check for commands and go back to sleep?

**Comment by ggtimtom** [comment]
I want to build two sensors:
A [rain meter](https://www.thingiverse.com/thing:13383) using a hallsensor an a magnet

A digital water meter: In my installed watermeter a little metalsheet is rotating. As a sensor i will use an geartoothsensor.

**Comment by AndunHH** [comment]
Thanks for the great work done here. As this whole feature is not jet finished, I'm hesitating to create a new issue. If this here is not the right spot, please give me a hint, where I should put it.

I succesfully used the example for the Switch and the on/off light with the given "light" example on a second seeed xiao esp32-c6 and with an ikea power outlet.

I tried to expand the example, because I don't want to bind a power outlet **manually after every reset**. I already found a [similar ticket for the espressif SDK.](https://github.com/espressif/esp-idf/issues/11904), that tells me, that this is already possible.

I changed the example to not wait in the setup() for a bound device. But then: nothing happens, because _is_bound = false in the follow up routings.
If I bind the first light again, the whole switch starts working and also the formerly bound power outlet is switching as expected. (= the newly bound light and the power outlet are switching together)

Therefore, I assume, that during initialization the state of _is_bound is not correctly read from the memory, right?
I'm was looking at https://github.com/espressif/arduino-esp32/blob/e99f1fdc6d12e07324be6e956318496e6a0bfac3/libraries/Zigbee/src/ZigbeeEP.cpp#L13 but at this point my knowdlege comes to an end. ...

Thanks for any hints how to solve this, or if this already known and planned. 
Or where I should open a individual ticket.

**Comment by Hedda** [comment]
@P-R-O-C-H-Y have you thought about maybe splitting Zigbee API for Arduino to make one part hardware-agnostic?

That is, breakout the Zigbee API from arduino-esp32 into its own separate library makes that hardware independent so that it could be reused by other projects with other Zigbee radio libraries, especially if they too have based their Zigbee stack on ZBOSS,  such as example Nordic Semiconductor. With the point being that such a library would attract more Arduino developers that are interested in this Zigbee API but might want to instead make an implementation for a non-ESP32 based radio.

**Comment by P-R-O-C-H-Y** [comment]
> This is great work, and i managed to get it to work immediately using a waveshare ESP32-C6 devkit board. I'm curious about the possiblities with sleep. One of the major reasons i'm using zigbee devices is when they need to run on low power. I have a bunch TRVs that run on battery at last a long time (months), yet respond to commands almost immediately (less than 5 seconds delay). My specific usecase is having a battery powered LED light that i only need to activate every now and then, but should not drain the battery when not in use. Zigbee would be perfect for this.
> 
> With deep sleep i'm getting a full reboot, and if i want reasonably quick response times i can only sleep for relatively short times. So i would assume i need some form of light sleep for that? Just wake up enough to check for commands and go back to sleep?

Hi @tlanfer, thanks for your feedback. 
The light sleep is not possible, as it needs the Power Management enabled and use the FreeRTOS tickless mode. We don't want to enable this settings for whole Arduino core, as it may bring unexpected behavior. If you really need the light sleep, you can use the ESP-IDF and the ESP-ZIGBEE-SDK directly. There is an example on how to use the light sleep :)

**Comment by P-R-O-C-H-Y** [comment]
> Thanks for the great work done here. As this whole feature is not jet finished, I'm hesitating to create a new issue. If this here is not the right spot, please give me a hint, where I should put it.
> 
> I succesfully used the example for the Switch and the on/off light with the given "light" example on a second seeed xiao esp32-c6 and with an ikea power outlet.
> 
> I tried to expand the example, because I don't want to bind a power outlet **manually after every reset**. I already found a [similar ticket for the espressif SDK.](https://github.com/espressif/esp-idf/issues/11904), that tells me, that this is already possible.
> 
> I changed the example to not wait in the setup() for a bound device. But then: nothing happens, because _is_bound = false in the follow up routings. If I bind the first light again, the whole switch starts working and also the formerly bound power outlet is switching as expected. (= the newly bound light and the power outlet are switching together)
> 
> Therefore, I assume, that during initialization the state of _is_bound is not correctly read from the memory, right? I'm was looking at
> 
> https://github.com/espressif/arduino-esp32/blob/e99f1fdc6d12e07324be6e956318496e6a0bfac3/libraries/Zigbee/src/ZigbeeEP.cpp#L13
> 
> but at this point my knowdlege comes to an end. ...
> Thanks for any hints how to solve this, or if this already known and planned. Or where I should open a individual ticket.

Hi @AndunHH, thank you for your feedback. I will take a look on this issue with the `_is_bound` if I can see anything wrong there. The TODO comment is there for me, as I might transfer all those variables  to not be a static if possible. 
If you have more info about that issue you have also with an example and steps how to reproduce it, please open a new issue and describe everything there. It will help a lot to have it separated, so we can discuss it there :)

**Comment by P-R-O-C-H-Y** [comment]
> @P-R-O-C-H-Y have you thought about maybe splitting Zigbee API for Arduino to make one part hardware-agnostic?
> 
> That is, breakout the Zigbee API from arduino-esp32 into its own separate library makes that hardware independent so that it could be reused by other projects with other Zigbee radio libraries, especially if they too have based their Zigbee stack on ZBOSS, such as example Nordic Semiconductor. With the point being that such a library would attract more Arduino developers that are interested in this Zigbee API but might want to instead make an implementation for a non-ESP32 based radio.

Hi @Hedda, I was not thinking about that as I use the ESP-ZIGBEE-SDK underneath, which uses the ZBOSS. The main idea of Arduino APIs is to offer a similar approach to other Arduino interfaces. If users want to use ESP-IDF Zigbee APIs, that is still possible, they can just ignore the Arduino Zigbee APIs. This is valid for most of our Arduino APIs, actually, users can always use the ESP-IDF underneath. We don't have a goal of portability of our libraries, only to keep portability of user applications. But for sure other manufacturers can follow the same APIs, but the Implementation will depend on them.

**Comment by HexiDev** [comment]
Hi! I'm not sure. But is there already a wrapper so you can communicate from device to device e.g. a button that toggles another zigbee light without using automations like homeassistant and/or nodered

**Comment by HexiDev** [comment]
@P-R-O-C-H-Y Hi! I'm not sure. But is there already a wrapper so you can communicate from device to device e.g. a button that toggles another zigbee light without using automations like homeassistant and/or nodered

**Comment by P-R-O-C-H-Y** [comment]
> @P-R-O-C-H-Y Hi! I'm not sure. But is there already a wrapper so you can communicate from device to device e.g. a button that toggles another zigbee light without using automations like homeassistant and/or nodered
> 
> 

Hi @HexiDev, just flash one ESP with on/off light and second ESP with on/off switch examples. The switch will act as a coordinator and create the Zigbee network, light will join the network and bind to the switch. So clicking a button on switch will turn on/off the light :)

**Comment by HexiDev** [comment]
But would it be also possible to use an existing network and make 2 zigbee end/router devices talk to eachother? Like in a sense that I can setup a switch as a router that can turn on/off my light that is also a router.

instead of communicating coordinator > end/router device

**Comment by Hedda** [comment]
@HexiDev the Zigbee function you are referring to is called Zigbee Binding, and you normally ”bind” two compatible cluster attributes. 

Binding is an on-device feature for Zigbee devices. It provides a mechanism for attaching an endpoint of one Zigbee device to an endpoint of another Zigbee device or to a Zigbee group.

For example, binding a “target destination” Zigbee device like a remote to a Zigbee light bulb, switch or group of light bulbs allows direct control of the “target” device (light, switch, shade) from the “remote” Zigbee device, bypassing the Zigbee Gateway. This means that the remote can control the light bulb or group even when any Zigbee Coordinator and Zigbee Gateway is not active.

Note that not all devices support binding. By default, most any Zigbee Gateway applications binds remotes to the coordinator, so click events are still forwarded to the Zigbee Gateway application for state reporting.

Note that as some remotes can only be bound to a single destination, you might need to unbind the remote from the coordinator before binding it to another device or group.

See how bindings is used in other software:

* https://www.home-assistant.io/integrations/zha#zigbee-binding-and-unbinding

* https://www.zigbee2mqtt.io/guide/usage/binding.html

* https://www.openhab.org/addons/bindings/zigbee/

**Comment by patrickfnielsen** [comment]
Examples/support for illuminance/occupancy would be nice, or maybe an example for how support could be implemented if you know the clusters/endpoints/attributes needed.

**Comment by lsroka76** [comment]
Does anybody successfully compiled Zigbee Gateway example in Arduino. Especially ot_rcp part?

**Comment by Farisabdurrasid** [comment]
@P-R-O-C-H-Y hi, can you help me?, I'm not sure if there is a way to configure esp32c6 as an end device and esp32H2 as a coordinator so they can connect to each other by sending messages via serial monitor. Is it possible and may I know where I can get the code?

**Comment by P-R-O-C-H-Y** [comment]
> @P-R-O-C-H-Y hi, can you help me?, I'm not sure if there is a way to configure esp32c6 as an end device and esp32H2 as a coordinator so they can connect to each other by sending messages via serial monitor. Is it possible and may I know where I can get the code?

Hi @Farisabdurrasid. To get this correctly, you want to use Zigbee protocol to send a messages between 2 devices (coordinator <-> end device) via serial monitor? Idk if this is the purpose of the Zigbee library and overall the Zigbee HA devices. What kind of messages it should be? Or can you explain it more deeply? Thanks

**Comment by Farisabdurrasid** [comment]
yes like that, without having to use Home Assistant, so these 2 ESP32C6
devices communicate with each other using zigbee.

For the project I am planning to create, the ESP32-C6 (coordinator) will
act as the MCU for the soil moisture sensor. When the soil moisture level
reaches a set point, another ESP32-C6 (end device) will act as the MCU to
change the traffic light's color. The data from the coordinator will be
sent to the end device using Zigbee communication. can it?

Pada tanggal Rab, 8 Jan 2025 15.55, Jan Procházka ***@***.***>
menulis:

**Comment by Farisabdurrasid** [comment]
> > @P-R-O-C-H-Y hi, can you help me?, I'm not sure if there is a way to configure esp32c6 as an end device and esp32H2 as a coordinator so they can connect to each other by sending messages via serial monitor. Is it possible and may I know where I can get the code?
> 
> Hi @Farisabdurrasid. To get this correctly, you want to use Zigbee protocol to send a messages between 2 devices (coordinator <-> end device) via serial monitor? Idk if this is the purpose of the Zigbee library and overall the Zigbee HA devices. What kind of messages it should be? Or can you explain it more deeply? Thanks

@P-R-O-C-H-Y yes like that, without having to use Home Assistant, so these 2 ESP32C6 devices communicate with each other using zigbee.

For the project I am planning to create, the ESP32-C6 (coordinator) will act as the MCU for the soil moisture sensor. When the soil moisture level reaches a set point, another ESP32-C6 (end device) will act as the MCU to change the traffic light's color. The data from the coordinator will be sent to the end device using Zigbee communication. can it?

**Comment by teambvd** [comment]
> For the project I am planning to create, the ESP32-C6 (coordinator) will act as the MCU for the soil moisture sensor. When the soil moisture level reaches a set point, another ESP32-C6 (end device) will act as the MCU to change the traffic light's color. The data from the coordinator will be sent to the end device using Zigbee communication. can it?

Sounds like what you're looking for is something more like ESP-NOW. Any to any comms isn't really what ZigBee does, and trying to do it by working within the ZigBee protocol is just going to unnecessarily complicate implementing what you're looking for IMO.

**Comment by Farisabdurrasid** [comment]
Yes, it makes sense.

Thanks for the knowledge

Pada tanggal Kam, 23 Jan 2025 23.42, teambvd ***@***.***>
menulis:

**Comment by MegaJack22** [comment]
Hey PROCHY,
First off, I just wanted to say a huge thanks for everything you’ve done for the community. Your work on this project is absolutely amazing, and it’s clear how much effort and time you’ve put into it. It’s inspiring to see someone so dedicated to making things better for everyone else. Your contributions have helped so many people, including me, to build some really cool stuff.
Now, I had a quick question for you. Do you think it would be possible to implement Zigbee OTA updates through systems like Z2M or similar platforms? This could add even more flexibility to projects built using your library, allowing remote firmware updates without needing physical access to devices. What are your thoughts on that?
Thanks again for all your hard work—it truly makes a difference!

**Comment by P-R-O-C-H-Y** [comment]
Hi everyone, can you please try #10946 if the OTA works for you? Thanks

**Comment by quentains** [comment]
Wow. I bought an ESP32-H2 8 months ago to try out to build Zigbee sensors. I had to deal with the ESP-IDF documentation for several days before finally giving up. Now I'm back to the news and discovering this new arduino library. Thank you very much for your great work. It will clearly help the most beginners among us to get started! Has the PR about documentation already been created?

**Comment by P-R-O-C-H-Y** [comment]
Hi @quentains, the documentation is unfortunately not ready yet. I have a quick draft, but as I am still working on adding new stuff, this is on lower priority now. Most of the functionality is demonstrated in the examples, so you can for now take a look at them :)

**Comment by jaccobezemer** [comment]
As others I also got stuck on using the ESP-IDF platform.
Thanks P-R-O-C-H-Y for giving us this Arduino library. I finally managed to create a Dimmable Light that integrates with Home Assistant ZHA.
However, I what to integrate it in Philips Hue.
I managed to change the trust center key but had to do this in the library. It would be nice if this would become possible in the API wrapper.

And while I managed to get the commissioning process with the Philips Hue bridge starting, my light does not show up in the Hue (no light found).
If someone managed to get this working with de Arduino Wrapper I'd love to hear about it.

**Comment by P-R-O-C-H-Y** [comment]
Hi @jaccobezemer, can you please give me more info on what exactly you needed to change, so I can add an API for that? If you prefer we can discuss it on our Discord channel.

**Comment by jaccobezemer** [comment]
> Hi [@jaccobezemer](https://github.com/jaccobezemer), can you please give me more info on what exactly you needed to change, so I can add an API for that? If you prefer we can discuss it on our Discord channel.

In ZigbeeCore.cpp I've added the following below Initialize Zigbee stack:
```
// Initialize Zigbee stack
  log_d("Initialize Zigbee stack");
  esp_zb_init(zb_cfg);

  // allow joining the Philips Hue network(s)
  esp_zb_enable_joining_to_distributed(true);
  uint8_t secret_zll_trust_center_key[] = {
    // The Philips HUE Trust Center Key
    0x81, 0x42, 0x86, 0x86, 0x5D, 0xC1, 0xC8, 0xB2, 0xC8, 0xCB, 0xC5, 0x2E, 0x5D, 0x65, 0xD1, 0xB8
  };
  esp_zb_secur_TC_standard_distributed_key_set(secret_zll_trust_center_key);

```
Maybe the above is possible with a public set method like .hueCompatible(true) or .setAlternateTCkey(<key>);

Below is the log showing the commissioning process with the Hue bridge. Unfortunately the light doesn't show up in the bridge.

```
.......ZBOSS: nwk/nwk_join.c:186   low lqi 20
ZBOSS: nwk/nwk_join.c:186   low lqi 0
ZBOSS: nwk/nwk_join.c:186   low lqi 0
ZBOSS: nwk/nwk_join.c:186   low lqi 0
ZBOSS: nwk/nwk_join.c:186   low lqi 0
ZBOSS: nwk/nwk_join.c:186   low lqi 0
ZBOSS: nwk/nwk_join.c:293   Will assoc to pan 0xe3e9 on channel 25, dev 0xea86, iface_id 0
.....ZBOSS: mac/mac_associate.c:448   received association response
ZBOSS: zdo/zdo_app_join.c:771   CONGRATULATIONS! joined status 0, addr 0x1d8e, iface_id 0, ch 25, rejoin 0
ZBOSS: zdo/zdo_app_join.c:773   xpanid b7:26:42:b2:99:b4:00:3e
ZBOSS: secur/zdo_secur.c:211   NWK key 5f:aa:6c:d2:66:4b:6a:a7:20:91:92:45:82:bc:78:7e
ZBOSS: commissioning/bdb/zdo_commissioning_bdb.c:549   COMMISSIONING_STOP: app signal 10 comm status 0
```

Here's some additional information and credits:
[https://wejn.org/2025/01/zigbee-hue-llo-world/](url)
[https://github.com/espressif/esp-zigbee-sdk/issues/358](url)

**Comment by jaccobezemer** [comment]
Hi P-R-O-C-H-Y,

This example works https://github.com/wejn/esp32-huello-world/tree/efb210dc42a2b8a11f1d66c69ec135010910055e

In that example setting the distributed key is just after `esp_zb_init(zb_cfg);`

When setting the distributed key in your wrapper also after esp_zb_init, then it is responding when the Hue bridge is open for joining but it isn't found by the bridge. On the other hand it is still able to join my HomeAssistant ZHA network.

So I suspect, in your wrapper, setting the distributed key should be done else ware in the code.

Do you have a tip?

**Comment by jaccobezemer** [comment]
Yeah, I manged to get it working with Philips Hue with the help of wejn's blog post (https://wejn.org/2025/01/zigbee-hue-llo-world).

These are the changes I've made to the dimmable light example. By the way, with these changes the light works with Philips Hue and Home Assistant ZHA.

**Changes to ZigbeeCore.cpp**

After esp_zb_init(zb_cfg); add the following code

```
  // allow joining the Philips Hue network(s)
  esp_zb_enable_joining_to_distributed(true);
  uint8_t secret_zll_trust_center_key[] = {
      0x81, 0x42, 0x86, 0x86, 0x5D, 0xC1, 0xC8, 0xB2, 0xC8, 0xCB, 0xC5, 0x2E, 0x5D, 0x65, 0xD1, 0xB8
  };
  esp_zb_secur_TC_standard_distributed_key_set(secret_zll_trust_center_key);
```

**Changes to ZigbeeDimmableLight.cpp**

Line 13: change app_device_version = 0 to 1
`_ep_config = {.endpoint = _endpoint, .app_profile_id = ESP_ZB_AF_HA_PROFILE_ID, .app_device_id = ESP_ZB_HA_DIMMABLE_LIGHT_DEVICE_ID, .app_device_version = 1};
`

The light can now be added to the Philips Hue Bridge 👍

Now we need to correct the off behavior of the Hue Bridge.

Add the following code after _cluster_list = zigbee_dimmable_light_clusters_create(&light_cfg); in
ZigbeeDimmableLight::ZigbeeDimmableLight(uint8_t endpoint) : ZigbeeEP(endpoint)

```
  // correct the off behavior  
  // https://github.com/espressif/esp-zigbee-sdk/issues/457#issuecomment-2426128314
  uint16_t on_off_on_time = 0;
  bool on_off_global_scene_control = 0;
  esp_zb_attribute_list_t *onoff_attr_list = esp_zb_cluster_list_get_cluster(_cluster_list, ESP_ZB_ZCL_CLUSTER_ID_ON_OFF, ESP_ZB_ZCL_CLUSTER_SERVER_ROLE);
  esp_zb_on_off_cluster_add_attr(onoff_attr_list, ESP_ZB_ZCL_ATTR_ON_OFF_ON_TIME, &on_off_on_time);
  esp_zb_on_off_cluster_add_attr(onoff_attr_list, ESP_ZB_ZCL_ATTR_ON_OFF_GLOBAL_SCENE_CONTROL,&on_off_global_scene_control);

```
Of course these changes should also be made to the on/off light and the dimmable color light.

**Comment by P-R-O-C-H-Y** [comment]
Thank @jaccobezemer for nice investigation and description :)
Would you like to add those changes yourself in a PR or should I handle it? :)

**Comment by jaccobezemer** [comment]
Hi P-R-O-C-H-Y,
I'm willing to make the changes myself but I'm a hobbyist and never did a PR. So I don't know what the do's and don'ts are.

That said, I think that with these change the API is not yet compatible enough with Philips HUE. I've noticed that overnight my light is off while the Hue bridge still thinks it on (and it should be on). 
Also when the light level is changed at the end device the bridge is not updated.
With ZHA this works oké.
So, compatibility with the Hue bridge still needs more work.

**Comment by jaccobezemer** [comment]
> > Hi P-R-O-C-H-Y, I'm willing to make the changes myself but I'm a hobbyist and never did a PR. So I don't know what the do's and don'ts are.
> > That said, I think that with these change the API is not yet compatible enough with Philips HUE. I've noticed that overnight my light is off while the Hue bridge still thinks it on (and it should be on). Also when the light level is changed at the end device the bridge is not updated. With ZHA this works oké. So, compatibility with the Hue bridge still needs more work.
> 
> Hi [@jaccobezemer](https://github.com/jaccobezemer), thank you for your additional feedback. I can add the API for adding the keys etc. but I don't have the Philips HUE Bridge, so I cannot test anything. And as you said its still not working 100 %, I am not sure if we should add the support before figuring all things to make it work reliably.

Hi P-R-O-C-H-Y, I agree with you. First we must figure out how to make it reliably. I will try to capture some traffic between the Bridge and an original Philip Hue Led bulb to see what is missing.

**Comment by jaccobezemer** [comment]
Hi, how can I enable logging? I build with Zigbee ED -Debug but I definitely don't see all messages.

**Comment by jaccobezemer** [comment]
When my dimmable light is joined to Home Assistant ZHA and calling `ZigbeeDimmableLight::setLight` updates Home Assistant. I see activity in Wireshark when `ZigbeeDimmableLight::setLight` is called.

Now when I join this end-point with exact the same firmware to the Philips HUE network all is working form the App point of view. But the App is not updated when setLight (`esp_zb_zcl_set_attribute_val`) is called and I see no activity in Wireshark.

Is there someone who can give a hint in the right direction?

**Comment by cpuchip** [comment]
I've only been loosely following, I do have a hue hub and an esp32-c6 devkit 1.2 on had that I could test with, finding time is the hard part.

I would start by looking at what's different about the pairing process. Did ZHA make any different calls out to the light that Hue didn't, like setting up automatic reporting, or binding? Some Zigbee coordinators setup automatic reporting where the light reports it's status after an action, and some will just call for Report Attributes regularly. Could you post a capture of the joining of the light and then subsequent activity? though if this is your personal zigbee network that would expose your network key to the internet which isn't ideal and we can look at other ways of trouble shooting.

I'm not expert, I've just had to trouble shoot my company's own zigbee networks against Hue and it's pretty eye opening when staring at sniffer logs all day.

**Comment by jaccobezemer** [comment]
> I've only been loosely following, I do have a hue hub and an esp32-c6 devkit 1.2 on had that I could test with, finding time is the hard part.
> 
> I would start by looking at what's different about the pairing process. Did ZHA make any different calls out to the light that Hue didn't, like setting up automatic reporting, or binding? Some Zigbee coordinators setup automatic reporting where the light reports it's status after an action, and some will just call for Report Attributes regularly. Could you post a capture of the joining of the light and then subsequent activity? though if this is your personal zigbee network that would expose your network key to the internet which isn't ideal and we can look at other ways of trouble shooting.
> 
> I'm not expert, I've just had to trouble shoot my company's own zigbee networks against Hue and it's pretty eye opening when staring at sniffer logs all day.

Hi cpuchip, thank you for offering to help.

I've captured the commisioning and rejoining of my dimmable light (zbee_nwk.addr == 0xe0b6).
See below the capture file.

[zigbee capture dimmable light commisioning and rejoining.zip](https://github.com/user-attachments/files/18950020/zigbee.capture.dimmable.light.commisioning.and.rejoining.zip)

**Comment by VAVorontsov** [comment]
Hello @P-R-O-C-H-Y and everyone,

Thank you for your recent work on creating a high level wrapper for the Zigbee library.
It is now much easier to use on ESP32-C6 chips in Arduino that what was initially available.

Would it be possible to implement an endpoint with the PM2.5 measurement?
Presumably this would be similar to the CO2 measurement endpoint which I assume uses the Concentration measurement cluster (looking at ZCL V7 specification page 349 (4-51)).
Is there a possibility of having a more flexible end point that can report some of the other substances in the ZCL spec?

Are there plans to have similar voltage, current and power measurement endpoints and battery level reporting in the future? These would be highly useful for many projects.

Once again. Huge thanks for your work.

**Comment by P-R-O-C-H-Y** [comment]
@jaccobezemer What may be the thing HA is doing differently is the data pooling. 
You can quickly check this HA discussion: https://community.home-assistant.io/t/zigbee-zha-enable-polling-for-updates/717592
![](https://community-assets.home-assistant.io/original/4X/3/7/4/374d3a49aece6396f1161d0cb1fb4afb53c40bd0.png)

**Comment by P-R-O-C-H-Y** [comment]
> Hello [@P-R-O-C-H-Y](https://github.com/P-R-O-C-H-Y) and everyone,
> 
> Thank you for your recent work on creating a high level wrapper for the Zigbee library. It is now much easier to use on ESP32-C6 chips in Arduino that what was initially available.
> 
> Would it be possible to implement an endpoint with the PM2.5 measurement? Presumably this would be similar to the CO2 measurement endpoint which I assume uses the Concentration measurement cluster (looking at ZCL V7 specification page 349 (4-51)). Is there a possibility of having a more flexible end point that can report some of the other substances in the ZCL spec?
> 
> Are there plans to have similar voltage, current and power measurement endpoints and battery level reporting in the future? These would be highly useful for many projects.
> 
> Once again. Huge thanks for your work.

Hello @VAVorontsov, 

Yes I can add the PM2.5 measurement endpoint. Please open a Feature Request so its easier to track and plan the work :)
What other substances you would like to see? (you can also specify it in the feature request.

About the power management, currently only power source, and battery level are supported. This can be extended to also support current, voltage etc. 
These are currently available functions:
```cpp
typedef enum {
  ZB_POWER_SOURCE_UNKNOWN = 0x00,
  ZB_POWER_SOURCE_MAINS = 0x01,
  ZB_POWER_SOURCE_BATTERY = 0x03,
} zb_power_source_t;

// Set Power source and battery percentage for battery powered devices
void setPowerSource(zb_power_source_t power_source, uint8_t percentage);
void setBatteryPercentage(uint8_t percentage);
void reportBatteryPercentage();
```

**Comment by jaccobezemer** [comment]
> [@jaccobezemer](https://github.com/jaccobezemer) What may be the thing HA is doing differently is the data pooling. You can quickly check this HA discussion: https://community.home-assistant.io/t/zigbee-zha-enable-polling-for-updates/717592 ![](https://camo.githubusercontent.com/34288af44bbee73669fe2d7cfef94bb76671c8a966cb73f5ef66f74e615aa381/68747470733a2f2f636f6d6d756e6974792d6173736574732e686f6d652d617373697374616e742e696f2f6f726967696e616c2f34582f332f372f342f333734643361343961656365363339366631313631643063623166623461666235336334306264302e706e67)

Thanks Jan, but I don't have that problem with ZHA. It's working oké there. It's with the Philips Hue bridge where it doesn't work.

**Comment by P-R-O-C-H-Y** [comment]
> > [@jaccobezemer](https://github.com/jaccobezemer) What may be the thing HA is doing differently is the data pooling.
> 
> Thanks Jan, but I don't have that problem with ZHA. It's working oké there. It's with the Philips Hue bridge where it doesn't work.

Sure, but I pointed out what may be different on Philips Hue. If there is also any settings like that to enable the polling for updates. That may solve the problem.

**Comment by cpuchip** [comment]
It took an embarrassing long time to get your example working, mostly because I didn't have a good way of "factory" resetting the esp32-c6 so that it would network steer and pair to my hue bridge.

I have it paired with Hue and can see that it's pulling the esp32-c6 for on/off (cluster 6) and dimmable (cluster 8) but cluster 300 (color) has some unsupported it's returning, like hue and saturation.

![Image](https://github.com/user-attachments/assets/5c453ffd-a3ab-4b0b-a3f0-f9b1d4e1bea7)

![Image](https://github.com/user-attachments/assets/3a7ef521-0715-4c1c-8c53-5ff44249fa06)

Now Hue is asking for is: 4000, 0001, 0003, 0004, 0008, 4002
that's EnhancedCurrentHue, CurrentSaturation, CurrentX, CurrentY, ColorMode, ColorLoopActive

to which the ESP32-C6 says 4000, 0001, and 4002 are UNSUPPORTED
which are EnhancedCurrentHue, CurrentSaturation, ColorLoopActive respectively.

That may be where we're running into issues? ColorMode is set to 0x01 which is CurrentX/CurrentY

A future note to myself, mainly, the pairing key needed is the ZLL Trust Center key.

Note: short address from attached packet capture
31FA : Hue Bridge
1EAD : Hue temp bulb
8097 : ESP32-C6 color dim example 
[capture-2025_Feb_28-13_10_espc6_joined_locations.zip](https://github.com/user-attachments/files/19032194/capture-2025_Feb_28-13_10_espc6_joined_locations.zip)

**Comment by cpuchip** [comment]
@jaccobezemer 

now I don't know how to call ZigbeeDimmableLight::setLight or where to call it. so I am blocked on reproducing your issue at this point. Though figuring out how to implement those attributes would probably solve the problem would be my guess?

Also Hue automatically read attributes for on/off, dimm level, and color automatically and regularly does it every 3 seconds (at least when there's only 2 lights attached) Hue is very chatty Zigbee wise.

**Comment by jaccobezemer** [comment]
@cpuchip Many thanks for your efforts and insights.

Setlight is available in the arduino-esp32 dimmable light example.
By the way, I'm developing a dimmable light not a color dimmable light. So the issues with the color cluster is not the problem with my end-point.
What I noticed, like you, is that the Philips Bridge automatically reads the light level and on/off when using a Philips Light or another Zigbee 3.0 dimmable light but not with the Expressif implementation.
I hope we found out why that is.
Could it because of this? https://github.com/espressif/esp-zigbee-sdk/issues/580
Or maybe Philips expects that a mains powered light is also a router?
I'm also researching how to change these values.
![Image](https://github.com/user-attachments/assets/804a0993-7e9d-49df-b627-37b7f3de7a41)

**Comment by cpuchip** [comment]
@jaccobezemer My total bad, I went through and re read everything and I understand where I went wrong. I thought the blog post and linked repo was your example, not one that you were basing off of to get your dimmable example working. I'll push on that thread and make the changes you've made to the default example and see if I can get that working too.

I learned a lot in the process :D like how to build https://github.com/wejn/esp32-huello-world/tree/24d50e985e004b97d97231551daf6966734a7223 on windows. I also figured out how to get each of the esp-zigbee-sdk examples working as well on my esp32-c6 dev kit. It might take me a few days to find another time slot though.

I apologize for convoluting this thread a little.

**Comment by jaccobezemer** [comment]
@cpuchip. No need to apologize! I'm happy with your attention. 
I will wait patiently and continue with my own research in the meantime.

**Comment by cpuchip** [comment]
@jaccobezemer I was able to make the changes you suggested against the Arduino Zigbee library and I did get my esp32-c6 to join Hue hub once, and noticed that hue queried the endpoints, and but didn't regularly then go on to read the attributes regularly. The only thing I can think of is that the Hue bridge doesn't like something in the Node Description Response coming back from the esp32 or something in the Simple Description Response.

The only differences I found in the Node Description Response to a temp dimmable Hue light was
Hue:
Node Descriptor Info: 0x4001 (Router)
Capability : 0x8E (full functioning device)
Max Buffer Size was 0x52
Max Incoming Transfer Size is 0x0080
Max Outgoing Transfer Size is 0x0080

ESP32-C6
Node Descriptor Info: 0x4002 (End Device)
Capability: 0x8C (Reduced functioning device)
Max Buffer Size was 0x6C
Max Incoming Transfer Size is 0x064D
Max Outgoing Transfer Size is 0x064D

I don't think that's it really. I'll keep digging as I have more time.

Is there a Discord I could join and be a little more real time, I don't mind posting back here with major findings, I also don't mind adding creating a PR with any changes we come up with here that fixes this. I think it'd be fun actually.

Also thank you @P-R-O-C-H-Y I really like the fact that the button is dual purpose and can reset the esp32 on long press and short press does something, like change level. Nice touch with the example code!

**Comment by jaccobezemer** [comment]
@cpuchip It's nice to read that you experience the same issues as I am.
I'am not aware of a Discord. A PR is no problem. Would indeed be fun.

**Comment by sivar2311** [comment]
How about the ["Arduino Core for Espressif" Discord](https://discord.gg/mnh2Rrbe)?

**Comment by Hedda** [comment]
@P-R-O-C-H-Y Would you consider also adding Zigbee End Device support to the [ESPHome](https://esphome.io/) firmware for ESP32-H2 (and ESP32-C6) in a similar approach that [tomaszduda23](https://github.com/tomaszduda23) is now working on for a competing platform? Please see:

- https://github.com/esphome/esphome/pull/7340

Maybe you could collaborate with [tomaszduda23](https://github.com/tomaszduda23) or at least give feedback to hin on that pull request for ESPHome so that it can be implemented in a way that could also make it compatible with your/Espressif API for Zigbee within Arduino?

**Comment by jaccobezemer** [comment]
@cpuchip and @P-R-O-C-H-Y last update about my Philips Hue issues before changing to Discord.

I managed to get things going by defining the dimmable light also as a ZIGBEE_ROUTER. That changed it to a Full-Function device.

ZigBee Device Profile, Device Announcement, Nwk Addr: 0xf520, Ext Addr: Espressif_ff:fe:57:9e:8c
    Sequence Number: 1
    Nwk Addr of Interest: 0xf520
    Extended Address: Espressif_ff:fe:57:9e:8c (40:4c:ca:ff:fe:57:9e:8c)
    Capability Information: 0x8e
        .... ...0 = Alternate Coordinator: False
        .... ..1. = Full-Function Device: True
        .... .1.. = AC Power: True
        .... 1... = Rx On When Idle: True
        .0.. .... = Security Capability: False
        1... .... = Allocate Short Address: True
 
The Philips Hue bridge apparently expects that a light is also a router because it now reads the onoff and level attributes every 53-55 seconds.
That's what I was looking for 👌
Also changed the power source to `#define ESP_ZB_ZCL_BASIC_POWER_SOURCE_MAINS_SINGLE_FASE ((uint8_t)0x01)`
And the maximum level to 0xfe instead of 0xff. 0xff is ignored by the Hue bridge. So resetting to device when the maximum level was 0xff, didn't show in the Philips App.

What made this search more difficult than needed 😉

1. Tried to build the light as router but that didn't succeed because I had debugging on and my Devkit had just 4MB Flash. Didn't realize that if I had turned debugging off then 4MB flash would have been enough.
2. Ordered a DevKit with 8MB Flash, flashed the Dimmable Light as Router (with debugging on) and it started working as expected.
3. Realized that if I could/should turn of debugging it would probably fit on a 4MB device...duh

So, now it works as other Zigbee 3 devices with Philips Hue. Finally 😊🙌

Thanks @cpuchip for thinking along and thanks @P-R-O-C-H-Y for your great API wrapper.

**Comment by P-R-O-C-H-Y** [comment]
@jaccobezemer I am glad that its finally working for you :) 
Just a quick info about the power source, as it's already defined in enumeration in `ZigbeeEP.h`. So There is no need for additional definition :)
```
typedef enum {
  ZB_POWER_SOURCE_UNKNOWN = 0x00,
  ZB_POWER_SOURCE_MAINS = 0x01,
  ZB_POWER_SOURCE_BATTERY = 0x03,
} zb_power_source_t;
```

Is there anything that needs to be added in order for Philips Hue to work or it's just about the correct configuration?

**Comment by cpuchip** [comment]
@jaccobezemer After a bunch of testing and comparing with Huello-world example : https://github.com/wejn/esp32-huello-world/tree/24d50e985e004b97d97231551daf6966734a7223 which does work with the polling of the Hue bridge, I was suspecting it was that Huello world is a FFD (full functioning device with router capable zigbee radio firmware) and what arduino defaults to which is RFD (Reduced functioning device or end device firmware) I don't know why Hue wouldn't think to pull a device still if it's marked as mains powered....

I finally figured out how to switch the dimmable example over to ZCZR firmware to act as a full functioning device, and got that paired to the Hue bridge and indeed it pulls the light every 1 second or so.

as an aside I did get the bulb paired to my companies Zigbee bridge which does setup reporting on attribute change and have verified that that does work with the example arduino dimmable firmware, when clicking the boot button, the esp32-c6 will report level changes. Our bridge does not pull by using ReadAttribueRequests..... which is a long story. Anyways, there's something else missing in the pairing process that our Zigbee gateway is looking for that this example code isn't providing. I'll save that investigation for another day.

The Hue bridge does not setup binding or reporting schedule so the arduino example firmware will not automatically push it's attribute changes up, that's just how Hue works.

As a note @P-R-O-C-H-Y the only thing we need are
The additions of the correct off behavior
and updating .app_device_version = 1
`
ZigbeeDimmableLight::ZigbeeDimmableLight(uint8_t endpoint) : ZigbeeEP(endpoint) {
  _device_id = ESP_ZB_HA_DIMMABLE_LIGHT_DEVICE_ID;

  zigbee_dimmable_light_cfg_t light_cfg = ZIGBEE_DEFAULT_DIMMABLE_LIGHT_CONFIG();
  _cluster_list = zigbee_dimmable_light_clusters_create(&light_cfg);

  // correct the off behavior  
  // https://github.com/espressif/esp-zigbee-sdk/issues/457#issuecomment-2426128314
  uint16_t on_off_on_time = 0;
  bool on_off_global_scene_control = 0;
  esp_zb_attribute_list_t *onoff_attr_list = esp_zb_cluster_list_get_cluster(_cluster_list, ESP_ZB_ZCL_CLUSTER_ID_ON_OFF, ESP_ZB_ZCL_CLUSTER_SERVER_ROLE);
  esp_zb_on_off_cluster_add_attr(onoff_attr_list, ESP_ZB_ZCL_ATTR_ON_OFF_ON_TIME, &on_off_on_time);
  esp_zb_on_off_cluster_add_attr(onoff_attr_list, ESP_ZB_ZCL_ATTR_ON_OFF_GLOBAL_SCENE_CONTROL,&on_off_global_scene_control);
  // .

  _ep_config = {.endpoint = _endpoint, .app_profile_id = ESP_ZB_AF_HA_PROFILE_ID, .app_device_id = ESP_ZB_HA_DIMMABLE_LIGHT_DEVICE_ID, .app_device_version = 1};

  // set default values
  _current_state = false;
  _current_level = 255;
}
`

And the ability to set the ZLL truct center key in ZigbeeCore.cpp
`
  // allow joining the Philips Hue network(s)
  esp_zb_enable_joining_to_distributed(true);
  uint8_t secret_zll_trust_center_key[] = {
    // The Philips HUE Trust Center Key
    0x81, ...,  0xB8
  };
  esp_zb_secur_TC_standard_distributed_key_set(secret_zll_trust_center_key);
`

I would also suggest changes to the Zigbee_Dimmable_light.ino and the color one too

`
#if !defined(ZIGBEE_MODE_ED) && !defined(ZIGBEE_MODE_ZCZR)
#error "Zigbee device mode is not selected in Tools->Zigbee mode"
#endif

#include "Zigbee.h"

#ifdef ZIGBEE_MODE_ZCZR
zigbee_role_t role = ZIGBEE_ROUTER;  // or can be ZIGBEE_COORDINATOR
#else
zigbee_role_t role = ZIGBEE_END_DEVICE;
#endif
`

update role
`
// When all EPs are registered, start Zigbee in End Device mode
  if (!Zigbee.begin(role)) {
`

I can make an PR for that tomorrow, unless you just want to do it yourself which is fine, the only part I'm unsure on is a proper way to add an API for  `esp_zb_secur_TC_standard_distributed_key_set` before stack startup.

**Comment by jaccobezemer** [comment]
@cpuchip We've figured it out. See my post of yesterday where I came up with the same conclusion😊

As for the app_device_version, there is already a methode for that:
`zbDimmableLight.setVersion(1); // sets the app_device_version to 1, needed for Philips HUE`
And there is one for setting the powersource: `zbDimmableLight.setPowerSource(ZB_POWER_SOURCE_MAINS);`

What we mis so far @P-R-O-C-H-Y, is what @cpuchip mentioned in the previous post.
To sum it up:
We need a configurable option to set
`esp_zb_enable_joining_to_distributed(true);`
`esp_zb_secur_TC_standard_distributed_key_set(secret_zll_trust_center_key);`

Also we need a way to change an attribute value to change these attributes:
`esp_zb_on_off_cluster_add_attr(onoff_attr_list, ESP_ZB_ZCL_ATTR_ON_OFF_ON_TIME, &on_off_on_time);
esp_zb_on_off_cluster_add_attr(onoff_attr_list, ESP_ZB_ZCL_ATTR_ON_OFF_GLOBAL_SCENE_CONTROL,&on_off_global_scene_control);`

@cpuchip If you want to make a PR, please do. I'm happy to work on that with you.

**Comment by cpuchip** [comment]
I'm still working on this, I have a branch on my fork of this repo: https://github.com/cpuchip/arduino-esp32/tree/feature/zigbee-tc-support

Sorry it's taking me longer to get to this then I thought, but I'm making progress.

**Comment by tomfrenzel** [comment]
@P-R-O-C-H-Y I wanted to try out the OTA functionality but it was not clear to me how i should create the OTA binary. 
The image_builder_tool seems like it only works for esp-idf. But i want to create the OTA binary from my .ino file.

Is it even possible to do that?

**Comment by Ruedi177** [comment]
@P-R-O-C-H-Y Thank you so much for adding Zigbee support to the Arduine IDE! I have successfully connected my ESP32C6 to the Home Assistant using the Example script "Zigbee_Pressure_Flow_Sensor". I would like to adapt this script to report Power measurement to HA and the Pressure Sensor Value would work fine for this. The only problem I have is that I don't know how to change the unit of measurement from hPa to Watts. Would you be able to help me with this?
Thank you!

**Comment by vdovinmih** [comment]
Hi! @P-R-O-C-H-Y Thank you for your work on the library.

Does anyone know a way to implement an ESP device that connects to an existing network and can send on/off/toggle control commands directly to neighboring network devices?

I use the Home Assistant and want to implement direct binding functionality in my device to control power outlets. At the moment, I have implemented a custom device with parameters and sensors. All that remains is to teach him how to send commands.

**Comment by vdovinmih** [comment]
> Hi! [@P-R-O-C-H-Y](https://github.com/P-R-O-C-H-Y) Thank you for your work on the library.
> 
> Does anyone know a way to implement an ESP device that connects to an existing network and can send on/off/toggle control commands directly to neighboring network devices?

added the ZigbeeSwitch endpoint, leaving the entire project in endpoint mode. It is now possible to pair on the Home Assistant side. When pairing is set up, the endpoint does not receive any pairing events, and predictably cannot send a command to the power outlet. Once the device is rebooted, the ZigbeeSwitch class finds the connected device and can control it. I haven't figured out how to catch the right event. Still in the process.

**Comment by P-R-O-C-H-Y** [comment]
> [@P-R-O-C-H-Y](https://github.com/P-R-O-C-H-Y) Thank you so much for adding Zigbee support to the Arduine IDE! I have successfully connected my ESP32C6 to the Home Assistant using the Example script "Zigbee_Pressure_Flow_Sensor". I would like to adapt this script to report Power measurement to HA and the Pressure Sensor Value would work fine for this. The only problem I have is that I don't know how to change the unit of measurement from hPa to Watts. Would you be able to help me with this? Thank you!

Hi @Ruedi177, as the ZHA is reading an attribute from a specific cluster, in this case a pressure measurement cluster, it's expecting the value to be in hPa. I am worried you cannot change it easily. You may check some configuration files in HA if you will be able to do this or the better ask on HA forums or Discord. This is something you cannot do in the Zigbee library.
But also there exists a power measurement cluster, but it's not implemented yet. If you want to do that I would be glad to see it added to the Zigbee library. I can add this to my backlog also, but there are lot of ongoing tasks now so I won't get to that any time soon.

**Comment by tomfrenzel** [comment]
> Hi [@tomfrenzel](https://github.com/tomfrenzel), Its working with Arduino binaries too. After you compile code in Arduino IDE, you can click in the topper menu to Sketch -> Export compiled binaries or you can go to the folders manually and find them. For the `image_builder_tool` you need to use the ino.bin file as selected in the image.
> 
> <img alt="Image" width="1144" src="https://private-user-images.githubusercontent.com/90197375/426040889-465008b6-8b53-4015-aae9-c29633a1098f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI4NDM3MjMsIm5iZiI6MTc0Mjg0MzQyMywicGF0aCI6Ii85MDE5NzM3NS80MjYwNDA4ODktNDY1MDA4YjYtOGI1My00MDE1LWFhZTktYzI5NjMzYTEwOThmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzI0VDE5MTAyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2MWNiYTRkMjYzOGM3YWMzOTNiYTJiMjA0NjViMDdkMDgwM2Y0MTQ0YmQ5OWUxYTEyZWQ2NmUxOTA2Y2I5ZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WpH_MgHR6q-7MIKngtNxkN-84oY3npFUwP54HXACTtg">

Thanks for getting back to me! That seems to work.

But can you tell me how you tested the OTA functionallity? I tried it with Z2M but for some reason it won't show me the current FW version inside the state. The OTA Cluster is shown inside the overview tho. I also tried to add my FW to my OTA index but the logs from the ESP show that no matching update was found.

I'm kind of lost here :D Let me know if that is somesting to discuss in a seperate discussion/discord

**Comment by relenox** [comment]
I have multiple ZigbeeSwitch and ZigbeeLight endpoints, a Home Assistant coordinator, and a ZigbeeDisplay endpoint (a ZigbeeLight device with a display but without a bulb). Is there a way to send messages (such as the switch name as defined in Home Assistant and the time it was pressed) to the ZigbeeDisplay?

**Comment by hruba202** [comment]
Zdravím,skvělá prace,je super,že přibývají další příklady zigbee zařízeních.Moje znalosti arduina jsou hodně malé,žadné,tak bych se rád zeptal,jestli je na jednu esp desku možné zkombinovat více senzorů(bh1750+bme280) jako tomu je u wifi a případně pomoci s kodem.A na závěr ještě jednu otázku,je nějaká šance,že se objeví  příklad pro mmwave radar?Děkuji a ať se daří.

**Comment by Hedda** [comment]
@hruba202 this is an English speaking repository.
